src/Task/ScriptManager.php

```diff
@@ -34,26 +39,34 @@ public function getTasks($name, $stage = null)
     {
         $collect = function ($name) use (&$collect, $stage) {
             $task = $this->tasks->get($name);
-            if ($stage === null || $task->isForStages($stage)) {
-                if ($task instanceof GroupTask) {
-                    return array_map($collect, $task->getTasks());
-                } else {
-                    return array_merge(
-                        array_map($collect, $task->getBefore()),
-                        [$name],
-                        array_map($collect, $task->getAfter())
-                    );
-                }
-            } else {
+            if ($stage !== null && !$task->isForStages($stage)) {
                 return [];
             }
+
+            $relatedTasks = [];
+
+            if ($this->isHooksEnabled()) {
+                $relatedTasks = array_merge(array_map($collect, $task->getBefore()), $relatedTasks);
+            }
+
+            if ($task instanceof GroupTask) {
+                $relatedTasks = array_merge($relatedTasks, array_map($collect, $task->getGroup()));
+            } else {
+                $relatedTasks = array_merge($relatedTasks, [$task->getName()]);
+            }
+
+            if ($this->isHooksEnabled()) {
+                $relatedTasks = array_merge($relatedTasks, array_map($collect, $task->getAfter()));
+            }
+
+            return $relatedTasks;
         };
 
-        $script = $collect($name);
+        $taskHierarchy = $collect($name);
```

Comment:
@elfet I refactored it as non obvious name. `$script` doesn't reflect the contents as for me and `$taskHierarchy` does. `$collect` returns exactly hierarchal collection of all related tasks. If you want I will return previous name or rename to something better.